[ALLUXIO-1911] Separate initialization path for Configuration

apc999 opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-1911

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9648/
Test FAILed.

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -167,7 +157,7 @@ public Configuration() {
     mProperties.putAll(systemProps);
 
-    // Update alluxio.master_address based on if Zookeeper is used or not.
+    // Update alluxio.master.address based on if Zookeeper is used or not.
```
I think this comment should be removed completely, it is outdated.

jsimsa commented on core/common/src/main/java/alluxio/RuntimeConstants.java:
```diff
@@ -0,0 +1,30 @@
+/*
```
would it make sense to just make this part of the `Constants.java` file? 


the only place where `Constants.VERSION` is used is here so we could just not expose it (or adopt a convention that appends `_KEY` to all the property keys ... which I think would be useful anyhow)

apc999 replied:
it was in `Constants.java` before, but causing the problem of introducing circular dependency

apc999 replied:
The difference (I want to set after this PR) is, `Constants.java` serving constants determined in compile time, and `RuntimeConstants` serving constants determined once the process is launched.

jsimsa replied:
what was the circular dependency and how does it manifest in Jenkins?

would making Runtime a nested class of Constants work?

apc999 replied:
@aaudiber has more context on the bug (https://github.com/Alluxio/alluxio/pull/1905). basically, if code here moved to `Constants.java`, the static block of `Constants`and `Configuration` depends on each other, causing unstable build problem. Moving `RuntimeConstant` nested under `Constant` will not solve that problem

aaudiber replied:
The constructor of `Configuration` (called in the static block here) makes use of many fields of `Constants`. Moving this static initialization block to `Constants` will make it impossible to statically initialize `Constants` without constructing a `Configuration`, and impossible to construct a `Configuration` without having `Constants` be statically initialized. To avoid this problem, we should make sure that `Constants` can be statically initialized without depending on any other classes.

jsimsa replied:
thanks for the explanation ... on a related note, do you guys think it would make sense to move all the constants that are property key names to a separate class (e.g. PropertyKey), I find it mildly confusing that Constants contains both property key names and various other constants. I think `conf.get(PropertyKey.VERSION)` reads better than `conf.get(Constants.VERSION)` this also avoids confusion between `Constants.VERSION` and `RuntimeConstants.VERSION`

apc999 replied:
@jsimsa I support this idea. And I think it could be a separate enum Type to be more type safe.

jsimsa replied:
sgtm

apc999 replied:
but that probably should be in another PR. (A JIRA added ALLUXIO-1949).

jsimsa replied:
makes sense

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -104,11 +93,9 @@ public Configuration(Properties props) {
-   *
-   * Most clients will call this constructor to allow default loading of properties to happen.
    */
   public Configuration() {
```
should this be made private to force uses to use the factory methods?

apc999 replied:
agree. Let's see how far we can go in this PR. Too many references of this constructor (especially in test), and I don't want to further enlarge the footprint of this PR.

jsimsa replied:
I think there is actually value of replacing all the references in this PR ... that way we know that the default / worker / master / client configuration is used where it should be used

apc999 replied:
I removed the references to this `public constructor` in production code in this PR.
for test code, there are > 60 references, I think it might be better to have separate PR doing that.

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -70,32 +71,20 @@
-   * @param props override {@link Properties}
-   */
-  public Configuration(Properties props) {
+  public Configuration(Map<?, ?> props) {
```
should this be made private to force uses to use the factory methods?

apc999 replied:
same as above, this should be the goal and let's see how far we go in this PR (not having a too big PR)

jsimsa commented:
@apc999 thanks for the PR, did a pass and left some comments

jsimsa commented:
@apc999 thanks for the PR, did a pass and left some comments, also please fix the build failure

jsimsa commented on core/common/src/main/java/alluxio/RuntimeConstants.java:
```diff
@@ -0,0 +1,30 @@
+  /** The relative path to the Alluxio target jar. */
+  public static final String ALLUXIO_JAR =
+      "target/alluxio-" + VERSION + "-jar-with-dependencies.jar";
+
```
add

```
private RuntimeConstants() // prevent instantiation
```

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9718/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9719/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9729/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='org.alluxio:alluxio-core-common' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9729/org.alluxio$alluxio-core-common/testReport'>org.alluxio:alluxio-core-common</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9729/org.alluxio$alluxio-core-common/testReport/alluxio/ConfigurationTest/workerDefaultTest/'><strong>alluxio.ConfigurationTest.workerDefaultTest</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9729/org.alluxio$alluxio-core-common/testReport/alluxio/ConfigurationTest/masterDefaultTest/'><strong>alluxio.ConfigurationTest.masterDefaultTest</strong></a></li></ul>
Test FAILed.

jsimsa commented on core/common/src/main/java/alluxio/RuntimeConstants.java:
```diff
@@ -0,0 +1,32 @@
+  public static final String ALLUXIO_JAR =
+      "target/alluxio-" + VERSION + "-jar-with-dependencies.jar";
+
+  private RuntimeConstants() {
```
can this be just one line?

apc999 replied:
done

jsimsa commented:
one minor nit, once the failing tests are fix this LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9734/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9740/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9741/
Test PASSed.

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -518,4 +497,49 @@ private void checkUserFileBufferBytes() {
+     * @param conf the source configuration
+     * @return a new configuration with only properties copied from source
+     */
+    public static Configuration create(Configuration conf) {
```
should this be called clone?

apc999 replied:
updated.

apc999 replied:
after doing some search, I think implementing `clone` is not considered  a good practice compared to using copy factory method or constructor. more details can be found 
http://stackoverflow.com/questions/11905630/java-super-clone-method-and-inheritance

jsimsa replied:
sgtm, so you will use a copy constructor?

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -518,4 +497,49 @@ private void checkUserFileBufferBytes() {
+     * @param properties the source configuration
+     * @return a new configuration with only properties copied from source
+     */
+    public static Configuration create(Map<?, ?> properties) {
```
should this be called `fromProperties`?

apc999 replied:
done. And also moved from Factory class to object class---I think it is less ideal to have Factory here since we are not really exposing an interface.

jsimsa commented on examples/src/main/java/alluxio/examples/BasicNonByteBufferOperations.java:
```diff
@@ -123,4 +123,5 @@ private boolean read(FileSystem alluxioClient) throws IOException, AlluxioExcept
     }
     return true;
   }
+
```
nit: not needed

apc999 replied:
fixed

jsimsa commented:
@apc999 thanks, did another pass and left some minor comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9750/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9760/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9761/
Test FAILed.

jsimsa commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9773/
Test PASSed.

jsimsa commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -75,10 +75,18 @@
+   * @param conf the source configuration to copy
+   * @return a new configuration instance with the properties copied from source configuration
+   */
+  public static Configuration copyFrom(Configuration conf) {
```
OOC, why not use a copy constructor?

apc999 replied:
yeah, I think copy constructor is better. Changed.

calvinjia commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -55,47 +54,69 @@
   /** File to set default properties. */
   public static final String DEFAULT_PROPERTIES = "alluxio-default.properties";
-  /** File to set customized properties. */
+  /** File to set customized properties for Alluxio deployment. */
```
Does this mean both client and server? If so I think its clearer to just state it explicitly.

apc999 replied:
done

calvinjia commented on core/server/src/main/java/alluxio/cli/ValidateConf.java:
```diff
@@ -36,9 +36,10 @@
    * @return true if the validation succeeds, false otherwise
    */
-  public static boolean validate() {
+  public static boolean validate(Configuration configuration) {
```
Should we make it clear that the command now only validates the client configuration?

apc999 replied:
not really. this method is also invoked by the master and worker. On the main, it is doing client-side check.

calvinjia commented on core/server/src/main/java/alluxio/master/AlluxioMaster.java:
```diff
@@ -71,12 +71,13 @@
 
     // validate the conf
-    if (!ValidateConf.validate()) {
+    if (!ValidateConf.validate(Configuration.createServerConf())) {
```
Should we validate the client conf whenever we start up the client context as well?

apc999 replied:
I am concerned by dong such a change: because that may bring down client-side job, right?

calvinjia replied:
Yes, it could cause clients to exit unexpectedly, but that might be the correct behavior if the configuration is flawed? Otherwise the client behavior might be unintended.

apc999 replied:
I updated the PR so alluxio shell will check the client-side conf before running.
for client conf in general, I didn't check it we don't have a good entrance point to check it (as in master or worker or shell, we have the main to check). It is possible that we could have it in `createClientConf`, but then it will be strange why we check in `createClientConf` but not in `createServerConf`

calvinjia commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -55,47 +54,69 @@
+  /**
+   * @return the default configuration without loading site and system properties
+   */
+  public static Configuration createDefaultConf() {
```
Could you add a TODO for unit testing the new constructors?

apc999 replied:
that's in the followup PR (adding unit tests for those constructors and utils)
TODO added

calvinjia commented:
@apc999 left a few comments, looks good overall.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9827/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9827/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9827/org.alluxio$alluxio-tests/testReport/alluxio.master/AlluxioMasterRestApiTest/getConfigurationTest/'><strong>alluxio.master.AlluxioMasterRestApiTest.getConfigurationTest</strong></a></li></ul>
Test FAILed.

apc999 commented:
@calvinjia  PTAL thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9828/
Test PASSed.

calvinjia commented on core/server/src/main/java/alluxio/cli/ValidateConf.java:
```diff
@@ -103,10 +103,19 @@ public static boolean validate() {
+      System.out.println("Server configuration has invalid entries.");
     }
-    System.out.println("All configuration entries are valid.");
     System.exit(0);
```
Should we `exit(-1)` still if either of the configurations are invalid?

calvinjia commented:
@apc999 Looks good overall, one minor comment. For validating the client configuration, I don't feel strongly about it if you think it will cause more trouble than benefit, but since the configuration change will be nontrivial, it may be worth making sure users don't have invalid configurations.

apc999 commented:
@calvinjia I updated the PR a bunch to clean up the configuration check . especially, I added check to alluxio shell too. I explained the reason it is not trivial to check client side in general. We could discuss it for sure.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9852/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9854/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9889/
Test PASSed.

calvinjia commented:
LGTM

apc999 commented:
thanks for the review. I will submit followup PRs for unit tests later.
